Map-Reduce Operation fails, when another instance tries to form a cluster with an instance running a map reduce task

One instance is running a map reduce task, and another instance tries to form a cluster with it, using tcp-ip join. This fails both the instances.

Log in the instance running map-reduce task.
Apr 27, 2014 7:48:05 PM com.hazelcast.mapreduce.impl.operation.RequestPartitionProcessed
SEVERE: [146.193.41.146]:5701 [main] [3.2] java.util.concurrent.ExecutionException: java.lang.NullPointerException
java.lang.RuntimeException: java.util.concurrent.ExecutionException: java.lang.NullPointerException
    at com.hazelcast.mapreduce.impl.MapReduceUtil.executeOperation(MapReduceUtil.java:248)
    at com.hazelcast.mapreduce.impl.task.JobSupervisor.checkFullyProcessed(JobSupervisor.java:288)
    at com.hazelcast.mapreduce.impl.task.JobProcessInformationImpl.updatePartitionState(JobProcessInformationImpl.java:115)
    at com.hazelcast.mapreduce.impl.task.JobProcessInformationImpl.updatePartitionState(JobProcessInformationImpl.java:103)
    at com.hazelcast.mapreduce.impl.MapReduceUtil.compareAndSwapPartitionState(MapReduceUtil.java:146)
    at com.hazelcast.mapreduce.impl.MapReduceUtil.stateChangeReducing(MapReduceUtil.java:122)
    at com.hazelcast.mapreduce.impl.MapReduceUtil.stateChange(MapReduceUtil.java:106)
    at com.hazelcast.mapreduce.impl.operation.RequestPartitionProcessed.run(RequestPartitionProcessed.java:71)
    at com.hazelcast.spi.impl.BasicOperationService.processOperation(BasicOperationService.java:363)
    at com.hazelcast.spi.impl.BasicOperationService.runOperation(BasicOperationService.java:228)
    at com.hazelcast.spi.impl.BasicInvocation.doInvoke(BasicInvocation.java:370)
    at com.hazelcast.spi.impl.BasicInvocation.invoke(BasicInvocation.java:190)
    at com.hazelcast.spi.impl.BasicInvocationBuilder.invoke(BasicInvocationBuilder.java:48)
    at com.hazelcast.mapreduce.impl.MapReduceService.processRequest(MapReduceService.java:223)
    at com.hazelcast.mapreduce.impl.task.JobSupervisor.processReducerFinished0(JobSupervisor.java:427)
    at com.hazelcast.mapreduce.impl.task.JobSupervisor.access$000(JobSupervisor.java:64)
    at com.hazelcast.mapreduce.impl.task.JobSupervisor$1.run(JobSupervisor.java:405)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:471)
    at java.util.concurrent.FutureTask.run(FutureTask.java:262)
    at com.hazelcast.util.executor.CompletableFutureTask.run(CompletableFutureTask.java:57)
    at com.hazelcast.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:186)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:744)
    at com.hazelcast.util.executor.PoolExecutorThreadFactory$ManagedThread.run(PoolExecutorThreadFactory.java:59)
Caused by: java.util.concurrent.ExecutionException: java.lang.NullPointerException
    at com.hazelcast.spi.impl.BasicInvocation$InvocationFuture.resolveResponseOrThrowException(BasicInvocation.java:792)
    at com.hazelcast.spi.impl.BasicInvocation$InvocationFuture.get(BasicInvocation.java:696)
    at com.hazelcast.spi.impl.BasicInvocation$InvocationFuture.get(BasicInvocation.java:674)
    at com.hazelcast.mapreduce.impl.MapReduceUtil.executeOperation(MapReduceUtil.java:242)
    ... 24 more
Caused by: java.lang.NullPointerException
    at com.hazelcast.mapreduce.impl.operation.GetResultOperation.run(GetResultOperation.java:50)
    at com.hazelcast.spi.impl.BasicOperationService.processOperation(BasicOperationService.java:363)
    at com.hazelcast.spi.impl.BasicOperationService.processPacket(BasicOperationService.java:309)
    at com.hazelcast.spi.impl.BasicOperationService.access$400(BasicOperationService.java:102)
    at com.hazelcast.spi.impl.BasicOperationService$BasicOperationProcessorImpl.process(BasicOperationService.java:756)
    at com.hazelcast.spi.impl.BasicOperationScheduler$ProcessTask.run(BasicOperationScheduler.java:186)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:744)
    at com.hazelcast.util.executor.PoolExecutorThreadFactory$ManagedThread.run(PoolExecutorThreadFactory.java:59)
    at ------ End remote and begin local stack-trace ------.(Unknown Source)
    at com.hazelcast.spi.impl.BasicInvocation$InvocationFuture.resolveResponse(BasicInvocation.java:836)
    at com.hazelcast.spi.impl.BasicInvocation$InvocationFuture.resolveResponseOrThrowException(BasicInvocation.java:769)
    ... 27 more

---

Log in the second instance, that attempts to join:
Apr 27, 2014 7:50:09 PM com.hazelcast.mapreduce.impl.operation.GetResultOperation
SEVERE: [146.193.41.145]:5701 [main] [3.2] null
java.lang.NullPointerException
    at com.hazelcast.mapreduce.impl.operation.GetResultOperation.run(GetResultOperation.java:50)
    at com.hazelcast.spi.impl.BasicOperationService.processOperation(BasicOperationService.java:363)
    at com.hazelcast.spi.impl.BasicOperationService.processPacket(BasicOperationService.java:309)
    at com.hazelcast.spi.impl.BasicOperationService.access$400(BasicOperationService.java:102)
    at com.hazelcast.spi.impl.BasicOperationService$BasicOperationProcessorImpl.process(BasicOperationService.java:756)
    at com.hazelcast.spi.impl.BasicOperationScheduler$ProcessTask.run(BasicOperationScheduler.java:186)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:744)
    at com.hazelcast.util.executor.PoolExecutorThreadFactory$ManagedThread.run(PoolExecutorThreadFactory.java:59)

---

When the second node attempts to join the Hazelcast cluster, immediately after the map-reduce task is started, it gave the below error. Probably these two are unrelated issues.

Apr 27, 2014 7:41:00 PM com.hazelcast.mapreduce.impl.operation.FireNotificationOperation
SEVERE: [146.193.41.145]:5701 [main] [3.2] JobSupervisor name=default, jobId=c9cb055c-7d94-4ceb-a1cc-37735ea72cf8 not found
java.lang.NullPointerException: JobSupervisor name=default, jobId=c9cb055c-7d94-4ceb-a1cc-37735ea72cf8 not found
    at com.hazelcast.mapreduce.impl.MapReduceService.dispatchEvent(MapReduceService.java:250)
    at com.hazelcast.mapreduce.impl.operation.FireNotificationOperation.run(FireNotificationOperation.java:54)
    at com.hazelcast.spi.impl.BasicOperationService.processOperation(BasicOperationService.java:363)
    at com.hazelcast.spi.impl.BasicOperationService.access$300(BasicOperationService.java:102)
    at com.hazelcast.spi.impl.BasicOperationService$LocalOperationProcessor.run(BasicOperationService.java:777)
    at com.hazelcast.util.executor.CachedExecutorServiceDelegate$Worker.run(CachedExecutorServiceDelegate.java:186)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:744)
    at com.hazelcast.util.executor.PoolExecutorThreadFactory$ManagedThread.run(PoolExecutorThreadFactory.java:59)

Second instance does nothing except joining the cluster, and participating in the map-reduce tasks sent by the first instance. This worked if the instances formed the cluster, BEFORE the map-reduce execution started. except during the scenarios where #2105 is encountered.

These issues limit the usability of the Hazelcast-3.2/Map-Reduce. The map-reduce is just the provided word-count example, over 3 files of 6 - 8 MB.

--- a/hazelcast/src/main/java/com/hazelcast/mapreduce/impl/MapReduceUtil.java
+++ b/hazelcast/src/main/java/com/hazelcast/mapreduce/impl/MapReduceUtil.java
@@ -245,13 +245,19 @@ public final class MapReduceUtil {
                     operation.run();
 
                     if (returnsResponse) {
-                        results.add((V) operation.getResponse());
+                        V response = (V) operation.getResponse();
+                        if (response != null) {
+                            results.add(response);
+                        }
                     }
                 } else {
                     if (returnsResponse) {
                         InvocationBuilder ib = os.createInvocationBuilder(SERVICE_NAME, operation, member.getAddress());
 
-                        results.add((V) ib.invoke().get());
+                        V response = (V) ib.invoke().get();
+                        if (response != null) {
+                            results.add(response);
+                        }
                     } else {
                         os.send(operation, member.getAddress());
                     }

--- a/hazelcast/src/main/java/com/hazelcast/mapreduce/impl/operation/GetResultOperation.java
+++ b/hazelcast/src/main/java/com/hazelcast/mapreduce/impl/operation/GetResultOperation.java
@@ -45,17 +45,20 @@ public class GetResultOperation
     @Override
     public void run()
             throws Exception {
+
         MapReduceService mapReduceService = getService();
         JobSupervisor supervisor = mapReduceService.getJobSupervisor(getName(), getJobId());
-        result = supervisor.getJobResults();
+        if (supervisor != null) {
+            result = supervisor.getJobResults();
 
-        // This is the final call so cleanup on all nodes that are not job owners
-        if (!supervisor.isOwnerNode()) {
-            mapReduceService.destroyJobSupervisor(supervisor);
-            AbstractJobTracker jobTracker = (AbstractJobTracker) mapReduceService.getJobTracker(getName());
-            jobTracker.unregisterTrackableJob(getJobId());
-            jobTracker.unregisterMapCombineTask(getJobId());
-            jobTracker.unregisterReducerTask(getJobId());
+            // This is the final call so cleanup on all nodes that are not job owners
+            if (!supervisor.isOwnerNode()) {
+                mapReduceService.destroyJobSupervisor(supervisor);
+                AbstractJobTracker jobTracker = (AbstractJobTracker) mapReduceService.getJobTracker(getName());
+                jobTracker.unregisterTrackableJob(getJobId());
+                jobTracker.unregisterMapCombineTask(getJobId());
+                jobTracker.unregisterReducerTask(getJobId());
+            }
         }
     }
 

--- a/hazelcast/src/main/java/com/hazelcast/mapreduce/impl/task/JobSupervisor.java
+++ b/hazelcast/src/main/java/com/hazelcast/mapreduce/impl/task/JobSupervisor.java
@@ -286,8 +286,8 @@ public class JobSupervisor {
             String name = configuration.getName();
             String jobId = configuration.getJobId();
             NodeEngine nodeEngine = configuration.getNodeEngine();
-            List<Map> results = MapReduceUtil
-                    .executeOperation(new GetResultOperationFactory(name, jobId), mapReduceService, nodeEngine, true);
+            GetResultOperationFactory operationFactory = new GetResultOperationFactory(name, jobId);
+            List<Map> results = MapReduceUtil.executeOperation(operationFactory, mapReduceService, nodeEngine, true);
 
             boolean reducedResult = configuration.getReducerFactory() != null;
 

